src/audio/google_rtc_audio_processing.c

```diff
@@ -0,0 +1,385 @@
+
+	comp_dbg(dev, "google_rtc_audio_processing_free()");
+
+	GoogleRtcAudioProcessingFree(cd->state);
+	cd->state = NULL;
+	rfree(cd->output_buffer);
+	rfree(cd->aec_reference_buffer);
+	rfree(cd->raw_mic_buffer);
+	comp_data_blob_handler_free(cd->tuning_handler);
+	rfree(cd);
+	rfree(dev);
+}
+
+
+static int google_rtc_audio_processing_trigger(struct comp_dev *dev, int cmd)
+{
+	comp_dbg(dev, "google_rtc_audio_processing_trigger(): %d", cmd);
+
+	return comp_set_state(dev, cmd);
+}
+
+static int google_rtc_audio_processing_prepare(struct comp_dev *dev)
+{
+	struct google_rtc_audio_processing_comp_data *cd = comp_get_drvdata(dev);
+	struct comp_buffer *source_buffer;
+	struct list_item *source_buffer_list_item;
+	uint32_t flags;
+
+	comp_dbg(dev, "google_rtc_audio_processing_prepare()");
+
+	/* searching for stream and feedback source buffers */
+	list_for_item(source_buffer_list_item, &dev->bsource_list) {
+		source_buffer = container_of(source_buffer_list_item, struct comp_buffer,
+						 sink_list);
+		buffer_lock(source_buffer, &flags);
+		if (source_buffer->source->comp.type == SOF_COMP_DEMUX)
+			cd->aec_reference = source_buffer;
+		else if (source_buffer->source->comp.type == SOF_COMP_DAI)
+			cd->raw_microphone = source_buffer;
+		buffer_unlock(source_buffer, flags);
+	}
+
+	cd->output = list_first_item(&dev->bsink_list, struct comp_buffer, source_list);
+
+	switch (cd->output->stream.frame_fmt) {
+#if CONFIG_FORMAT_S16LE
+	case SOF_IPC_FRAME_S16_LE:
+		break;
+#endif /* CONFIG_FORMAT_S16LE */
+	default:
+		comp_err(dev, "unsupported data format: %d", cd->output->stream.frame_fmt);
+		return -EINVAL;
+	}
+
+	if (cd->output->stream.rate != GOOGLE_RTC_AUDIO_PROCESSING_SAMPLERATE) {
+		comp_err(dev, "unsupported samplerate: %d", cd->output->stream.rate);
+		return -EINVAL;
+	}
+
+	return comp_set_state(dev, COMP_TRIGGER_PREPARE);
+}
+
+static int google_rtc_audio_processing_reset(struct comp_dev *dev)
+{
+	comp_dbg(dev, "google_rtc_audio_processing_reset()");
+
+	return comp_set_state(dev, COMP_TRIGGER_RESET);
+}
+
+static int google_rtc_audio_processing_copy(struct comp_dev *dev)
+{
+	struct google_rtc_audio_processing_comp_data *cd = comp_get_drvdata(dev);
+
+	struct comp_copy_limits cl;
+	int frame;
+	uint32_t aec_reference_buff_frag;
+	uint32_t raw_microphone_buff_frag;
+	uint32_t output_buff_frag;
+	int16_t *src;
+	int16_t *dst;
+	uint32_t num_aec_reference_frames;
+	uint32_t num_aec_reference_bytes;
+	uint32_t flags;
+
+	buffer_lock(cd->aec_reference, &flags);
+	num_aec_reference_frames = audio_stream_get_avail_frames(&cd->aec_reference->stream);
+	num_aec_reference_bytes = audio_stream_get_avail_bytes(&cd->aec_reference->stream);
+	buffer_unlock(cd->aec_reference, flags);
+
+	buffer_invalidate(cd->aec_reference, num_aec_reference_bytes);
+
+	aec_reference_buff_frag = 0;
+	for (frame = 0; frame < num_aec_reference_frames; frame++) {
+		src = audio_stream_read_frag_s16(&cd->aec_reference->stream,
```

Comment:
Please avoid to read/write frag functions because they are slow, see https://github.com/thesofproject/sof/issues/4967

Reply:
I'll address that on `main` branch and then backport it.

Reply:
@singalsu are all the replacements on this branch already